Provide links to calendar in event creation/update activities

tcitworld opened the pull request:
Since Calendar 2.0 we have dedicated routes so we can link to events directly from the Activity app.

See also https://github.com/nextcloud/activity/pull/420 for proper formatting in the Activity app.

![Screenshot_2020-01-21 Activity - Nextcloud- 2](https://user-images.githubusercontent.com/2197836/72794070-aa880780-3c3b-11ea-9344-e0dec2ed5f8c.png)
(The last item has also a link to the event)

tcitworld committed:
Provide links to calendar in event creation/update activities

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld commented on apps/dav/lib/CalDAV/Activity/Backend.php:
```diff
@@ -442,27 +441,48 @@ public function onTouchCalendarObject($action, array $calendarData, array $share
 	}
 
+	private function checkIfCalendarAppIsActivatedAndCorrectVersion(): bool {
+		$appManager = \OC::$server->getAppManager();
```
Autoinject is probably better here.

nickvergessen commented on apps/dav/lib/CalDAV/Activity/Backend.php:
```diff
@@ -442,27 +441,48 @@ public function onTouchCalendarObject($action, array $calendarData, array $share
 
+	private function checkIfCalendarAppIsActivatedAndCorrectVersion(): bool {
+		$appManager = \OC::$server->getAppManager();
+		return $appManager->isEnabledForUser('calendar') && version_compare($appManager->getAppVersion('calendar'), '2.0.0') >= 0;
```
Can we solve the version requirement by only having compatible releases available for 19?
cc @georgehrke

georgehrke replied:
Hm, I would prefer not artificially forcing users to update the server.

nickvergessen replied:
The other way around, but this is already the case. So the version comparison can be removed.

georgehrke reviewed (dismissed):
~I'm blocking, to put this on hold for now. I'm worried I have to change the url format in order to fix https://github.com/nextcloud/calendar/issues/1894. Will look into that issue today.~

Never mind, it's yet again related to unencoded group names

rullzer commented:
@georgehrke yay or nay?

georgehrke commented:
@tcitworld can you remove the version check as suggested by @nickvergessen.

Then this is good to go

tcitworld commented:
Removed the version checked, properly injected dependencies, moved `generateObjectParameter` to `OCA\DAV\CalDAV\Activity\Provider\Event` because not needed in `OCA\DAV\CalDAV\Activity\Provider\Base` and improved tests, so please give a quick look again.

georgehrke reviewed (approved):
Code looks good

nickvergessen commented:
Activity PR is merged

nickvergessen commented on apps/dav/lib/CalDAV/Activity/Backend.php:
```diff
@@ -83,6 +76,19 @@ public function onCalendarUpdate(array $calendarData, array $shares, array $prop
+	 * @param IUserSession $userSession
+	 * @param IAppManager $appManager
+	 */
+	public function __construct(IActivityManager $activityManager, IGroupManager $groupManager, IUserSession $userSession, IAppManager $appManager) {
```
can we keep the constructor as first function?

And feel free to remove the docs when moving it as they are just duplicating the stronger type hints.

nickvergessen commented on apps/dav/lib/CalDAV/Activity/Backend.php:
```diff
@@ -441,23 +447,39 @@ public function onTouchCalendarObject($action, array $calendarData, array $share
+			];
+
+			if ($object['type'] === 'event' && strpos($action, Event::SUBJECT_OBJECT_DELETE) === false && $this->appManager->isEnabledForUser('calendar')) {
+				$objectId = base64_encode('/remote.php/dav/calendars/' . $owner . '/' . $calendarData['uri'] . '/' . $objectData['uri']);
```
could this all be done on rendering instead?
Just store owner, firstoccurence and uri in the array,
but to all the calculation on rendering. That would allow us to always link to the next occurance of an repeating event instead of the first one from years ago?

tcitworld replied:
> could this all be done on rendering instead?

Yes, will do.

> That would allow us to always link to the next occurance of an repeating event instead of the first one from years ago?

I don't see how can I get the next occurrence of a repeating event without directly parsing the raw ICS data and [expanding it](https://sabre.io/vobject/recurrence/). I only have the first and the last, not the recurring rule.

tcitworld commented on apps/dav/lib/CalDAV/Activity/Backend.php:
```diff
@@ -435,29 +435,41 @@ public function onTouchCalendarObject($action, array $calendarData, array $share
 
-		foreach ($users as $user) {
+		// Users for share can return the owner itself if the calendar is published
+		foreach (array_unique($users) as $user) {
```
`getUsersForShares` potentially returning the owner itself because of public links for calendar should be fixed somewhere else, this is a temporary fix. We need a proper `nc:` property to indicate that. cc @georgehrke

tcitworld commented on apps/dav/lib/CalDAV/Activity/Provider/Event.php:
```diff
@@ -44,29 +47,67 @@ class Event extends Base {
+				$link = [
+					'view' => 'dayGridMonth',
+					'timeRange' => $timeRange->format('Y-m-d'),
+					'mode' => 'sidebar',
```
This could get user's `skipPopover` setting to know if we use the popover or the sidebar, do you want that?

georgehrke replied:
We actually have a special route for this, but it's also hardcoded to sidebar at the moment.

I would just keep it as is and improve later on :)

georgehrke replied:
If you use that route, we could actually get rid of `firstoccurrence`

tcitworld replied:
I'm using `/dayGridMonth/now/edit/sidebar/$objectId/next` then.

tcitworld commented:
Slighty off-topic, it's kinda sad we can't properly detect event being renamed, it makes you think you updated two different events when you just renamed one twice.
![image](https://user-images.githubusercontent.com/2197836/79902869-7676db00-8412-11ea-91d6-04682ccccdcb.png)

nickvergessen commented on apps/dav/lib/CalDAV/Activity/Provider/Event.php:
```diff
@@ -44,29 +47,67 @@ class Event extends Base {
+	 * @param array $eventData
+	 * @return array
+	 */
+	protected function generateObjectParameter(array $eventData) {
```
This was in the base because it's also used in Todo.php and it makes it more obvious that it is shared.
But I guess it's fine eitherway.

tcitworld replied:
Yeah, my reasoning was that Calendar doesn't need it.

nickvergessen commented on apps/dav/lib/CalDAV/Activity/Provider/Event.php:
```diff
@@ -44,29 +47,67 @@ class Event extends Base {
+					'objectId' => $objectId,
+					'recurrenceId' => $linkData['firstoccurence']
+				];
+				$params['link'] = $this->url->linkToRouteAbsolute('calendar.view.indexview.timerange.edit', $link);
```
this will :boom: when calendar is disabled,
so maybe we return the params without the link (since it's optional anyway) in the exception catching?

georgehrke replied:
So just wrap in `isEnabledForUser`?

nickvergessen replied:
Yes

tcitworld replied:
Not sure I follow here, we set the link inside the `AppManager::isEnabledForUser` check already, and params without the link will be return at the end properly.

nickvergessen replied:
Ah, didnt see it as it was all at the end of this long if

rullzer reviewed (approved):
Lets do this
